### PR TITLE
Magic Blog milestone 13 : Authorization

### PR DESCRIPTION
**Pull request by Nemwel-Boniface**
## Hello reviewer(s)

This is @kosher9 and @Nemwel-Boniface :sunglasses: 

In this milestone we were able to:

- [x] Install CanCanCan in your project
- [x] Add a role column to the user's table
- [x] Implement A user can delete a post if it is theirs or if they have an admin role
- [x] Implement A user can delete a comment if it is theirs or if they have an admin role

In case you face any challenge creating the db run the following:

`rake db:drop`
`rake db:create`
`rake db:schema:load`
`rake db:migrate` 

Thank you for taking the time to review my project :pray:

### CODE REVIEW

File: app/models/ability.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  include CanCan::Ability
+
+  def initialize(user)
+    # can :read, [Post, Comment], author: user
+    return unless user.present?
+
+    can :destroy, [Post, Comment]
+    return unless user.role?
+
+    can :destroy, [Post, Comment]
```
Comment (RolandM99) [inline app/models/ability.rb]: - [ ] You didn't yet implement your authorization as expected. for now, even the user who is not the owner of a post had the authorization to delete a comment. Kindly, fix this issue. You can take a look at this code below to know how to implement it.
```ruby
 class Ability
 include CanCan::Ability

     def initialize(user)

           user ||= User.new # guest user (not logged in)
  
           if user.admin? 
           can :manage, :all
         else
           can :manage, Post, author: user
           can :manage, Comment, author: user
           can :read, :all
         end
     end
 end
```
Kindly, try to fix that issue.

**Review by RolandM99 — CHANGES_REQUESTED** [review]
Hi @kosher9, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

### To highlight 💯 
- Good commit message ✔️ 
- No Linter errors ✔️ 
- Follow Github flow ✔️ 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me on @Rolandm99 in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -1,4 +1,7 @@
 class PostsController < ApplicationController
+  # load_and_authorize_resource
```
Comment (rustam-tolipov) [inline app/controllers/posts_controller.rb]: - [x] Why did you comment load_and_authorize? To run your Ability, you need to have this.

File: app/controllers/posts_controller.rb [hunk 3]
```diff
@@ -1,4 +1,7 @@
...
+  # before_action :set_post, only: [:index, :show, :destroy]
```
Comment (rustam-tolipov) [inline app/controllers/posts_controller.rb]: - [x] You do not need to have set_post if you are using load and authorize.

File: app/models/ability.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
+class Ability
+  include CanCan::Ability
+
+  def initialize(_user)
+    # can :read, [Post, Comment], author: user
```
Comment (rustam-tolipov) [inline app/models/ability.rb]: - [x] Kindly remove unused commented codes.

File: app/views/posts/index.html.erb [hunk 5]
```diff
@@ -19,9 +19,13 @@
         <h5>Likes : <%= post.likes_counter %></h5>
       </div>
     <% end %>
+    <% if can? :destroy, post %>
+      <div><%= button_to 'Delete', user_post_path(@user.id, post.id), method: :delete %></div>
+<!--      <button>Button</button>-->
```
Comment (rustam-tolipov) [inline app/views/posts/index.html.erb]: - [x] Kindly remove unused commented code.

**Review by rustam-tolipov — CHANGES_REQUESTED** [review]
Hi @Nemwel-Boniface, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
_Your project is almost complete!_
_No Linters errors_


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Mihndim2020 — APPROVED** [review]
# STATUS: APPROVED!!!  🎉 🎉 🎉 :champagne: :champagne: :champagne:
Hi @kosher9, 

### Good job !!!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Clean/Organized code
:heavy_check_mark:  Good commit messages
:heavy_check_mark: Descriptive Readme file

Your project is complete! Great work on making the changes requested by the previous reviewer. There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 🎉 :champagne: :champagne: :champagne:

Congratulations-Memes-FEATURED

## Optional suggestions

_Check the comments under the review._

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

### N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**